database/database.go

```diff
@@ -0,0 +1,160 @@
+		staticDB     *mongo.Database
+		staticLogger *logrus.Logger
+	}
+
+	// DBCredentials is a helper struct that binds together all values needed for
+	// establishing a DB connection.
+	DBCredentials struct {
+		User     string
+		Password string
+		Host     string
+		Port     string
+	}
+)
+
+// New creates a new database connection.
+func New(ctx context.Context, creds DBCredentials, logger *logrus.Logger) (*DB, error) {
+	return NewCustomDB(ctx, dbName, creds, logger)
+}
+
+// NewCustomDB creates a new database connection to a database with a custom name.
+func NewCustomDB(ctx context.Context, dbName string, creds DBCredentials, logger *logrus.Logger) (*DB, error) {
+	if ctx == nil {
+		return nil, errors.New("invalid context provided")
+	}
+	if logger == nil {
+		return nil, errors.New("invalid logger provided")
+	}
+	c, err := mongo.NewClient(options.Client().ApplyURI(connectionString(creds)))
+	if err != nil {
+		return nil, errors.AddContext(err, "failed to create a new db client")
+	}
+	err = c.Connect(ctx)
+	if err != nil {
+		return nil, errors.AddContext(err, "failed to connect to db")
+	}
+	db := c.Database(dbName)
+	err = ensureDBSchema(ctx, db, logger)
+	if err != nil {
+		return nil, err
+	}
+	return &DB{
+		staticCtx:    ctx,
+		staticDB:     db,
+		staticLogger: logger,
+	}, nil
+}
+
+// Disconnect closes the connection to the database in an orderly fashion.
+func (db *DB) Disconnect(ctx context.Context) error {
+	return db.staticDB.Client().Disconnect(ctx)
+}
+
+// Ping sends a ping command to verify that the client can connect to the DB and
+// specifically to the primary.
+func (db *DB) Ping(ctx context.Context) error {
+	ctx2, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+	return db.staticDB.Client().Ping(ctx2, readpref.Primary())
+}
+
+// connectionString is a helper that returns a valid MongoDB connection string
+// based on the passed credentials and a set of constants. The connection string
+// is using the standalone approach because the service is supposed to talk to
+// the replica set only via the local node.
+// See https://docs.mongodb.com/manual/reference/connection-string/
+func connectionString(creds DBCredentials) string {
+	// There are some symbols in usernames and passwords that need to be escaped.
+	// See https://docs.mongodb.com/manual/reference/connection-string/#components
+	return fmt.Sprintf(
+		"mongodb://%s:%s@%s:%s/?compressors=%s&readPreference=%s&w=%s&wtimeoutMS=%s",
+		url.QueryEscape(creds.User),
+		url.QueryEscape(creds.Password),
+		creds.Host,
+		creds.Port,
+		mongoCompressors,
+		mongoReadPreference,
+		mongoWriteConcern,
+		mongoWriteConcernTimeout,
+	)
+}
+
+// ensureDBSchema checks that we have all collections and indexes we need and
+// creates them if needed.
+// See https://docs.mongodb.com/manual/indexes/
+// See https://docs.mongodb.com/manual/core/index-unique/
+func ensureDBSchema(ctx context.Context, db *mongo.Database, log *logrus.Logger) error {
```

Comment:
we might be at a point where we should seriously consider extracting all this database template code into a repo, that will also help us apply the same design choices because there are definitely 2-3 styles floating around now

Reply:
As we've already discussed, I'm on board with that. Let's discuss it in a separate place, though.